File: 108-Convert Sorted Array to Binary Search Tree/ConvertSortedArrayBST.kt
fun main() {
    val nums = intArrayOf(-10, -3, 0, 5, 9)

    ConvertSortedArrayBST.sortedArrayToBST(nums)
}

class TreeNode(var `val`: Int) {
    var left: TreeNode? = null
    var right: TreeNode? = null
}

object ConvertSortedArrayBST {
    //
    // First attempt 07 June 2024
    // Runtime: 169ms, 37.66MB
    //
    fun sortedArrayToBST(nums: IntArray): TreeNode? {
        if (nums.isEmpty()) {
            return null
        }

        val mid = nums.size / 2
        val node = TreeNode(nums[mid])

        node.left = sortedArrayToBST(nums.sliceArray(0 until mid))
        node.right = sortedArrayToBST(nums.sliceArray(mid + 1 until nums.size))

        return node
    }
}

File: 108-Convert Sorted Array to Binary Search Tree/convert_sorted_array_bst.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func main() {
	nums := []int{-10, -3, 0, 5, 9}
	sortedArrayToBST(nums)
}

func sortedArrayToBST(nums []int) *TreeNode {
	//
	// First attempt 07 June 2024 11:25
	// Runtime: 3ms, Memory: 3.59MB
	//
	if len(nums) == 0 {
		return nil
	}

	mid := len(nums) / 2
	node := &TreeNode{Val: nums[mid]}

	node.Left = sortedArrayToBST(nums[:mid])
	node.Right = sortedArrayToBST(nums[mid+1:])

	return node
}
